online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala

```diff
@@ -0,0 +1,128 @@
+      catalystUtil: PooledCatalystUtil
+  ): DerivationFunc = {
+    {
+      case (keys: Map[String, Any], values: Map[String, Any]) =>
+        reintroduceExceptions(catalystUtil.performSql(keys ++ values).orNull, values)
+    }
+  }
+
+  def buildDerivationFunction(
+      derivationsScala: List[Derivation],
+      keySchema: StructType,
+      baseValueSchema: StructType
+  ): DerivationFunc = {
+    if (derivationsScala.isEmpty) {
+      return { case (_, values: Map[String, Any]) => values }
+    } else if (derivationsScala.areDerivationsRenameOnly) {
+      buildRenameOnlyDerivationFunction(derivationsScala)
+    } else {
+      val catalystUtil = buildCatalystUtil(derivationsScala, keySchema, baseValueSchema)
+      buildDerivationFunctionWithSql(catalystUtil)
+    }
+  }
+
+  def applyDeriveFunc(
+      deriveFunc: DerivationFunc,
+      request: Request,
+      baseMap: Map[String, AnyRef]
+  ): Map[String, AnyRef] = {
+    val requestTs = request.atMillis.getOrElse(System.currentTimeMillis())
+    val requestDs = TsUtils.toStr(requestTs).substring(0, 10)
+    // used for derivation based on ts/ds
+    val tsDsMap: Map[String, AnyRef] = {
+      Map("ts" -> (requestTs).asInstanceOf[AnyRef], "ds" -> (requestDs).asInstanceOf[AnyRef])
+    }
+    val derivedMap: Map[String, AnyRef] = Try(
+      deriveFunc(request.keys, baseMap ++ tsDsMap)
+        .mapValues(_.asInstanceOf[AnyRef])
+        .toMap) match {
+      case Success(derivedMap) => derivedMap
+      case Failure(exception) => {
+        throw exception
+      }
+    }
+    val derivedMapCleaned = derivedMap -- tsDsMap.keys
+    derivedMapCleaned
+  }
+
+  def buildCatalystUtil(
+      derivationsScala: List[Derivation],
+      keySchema: StructType,
+      baseValueSchema: StructType
+  ): PooledCatalystUtil = {
+    val baseExpressions = if (derivationsScala.derivationsContainStar) {
+      baseValueSchema
+        .filterNot { derivationsScala.derivationExpressionSet contains _.name }
+        .map(sf => sf.name -> sf.name)
+    } else { Seq.empty }
+    val expressions = baseExpressions ++ derivationsScala.derivationsWithoutStar.map { d => d.name -> d.expression }
+    new PooledCatalystUtil(expressions, StructType("all", (keySchema ++ baseValueSchema).toArray ++ timeFields))
+  }
+
+  def buildDerivedFields(
+      derivationsScala: List[Derivation],
+      keySchema: StructType,
+      baseValueSchema: StructType
+  ): Seq[StructField] = {
+    if (derivationsScala.areDerivationsRenameOnly) {
+      val baseExpressions = if (derivationsScala.derivationsContainStar) {
+        baseValueSchema.filterNot { derivationsScala.derivationExpressionSet contains _.name }
+      } else {
+        Seq.empty
+      }
+      val expressions: Seq[StructField] = baseExpressions ++ derivationsScala.derivationsWithoutStar.map { d =>
+        {
+          if (baseValueSchema.typeOf(d.expression).isEmpty) {
+            throw new IllegalArgumentException(
```

Comment:
same as above. 
let's not throw here. We could use datadog's increment exception method to capture it.

Reply:
The exception here will be caught when we call the getJoinCodecs function, recorded in datadog, and then returned as join_codec_fetch_exception in the exception map https://sourcegraph.d.musta.ch/github.com/airbnb/chronon/-/blob/online/src/main/scala/ai/chronon/online/Fetcher.scala?L246